File: src/traffic_control_system_detection_cpp/scripts/image_processor_only_opencv.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/compressed_image.hpp>
#include <team_interfaces/msg/signs.hpp>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include <ament_index_cpp/get_package_share_directory.hpp>

// Set up the path to the model
const std::string model_path = ament_index_cpp::get_package_share_directory("sign_detection") + "/model_files/best.onnx";

const std::vector<std::string> classes = {"cross_parking", "overtaking_allowed", "overtaking_forbidden",
                                          "parallel_parking", "pit_in", "pit_out"};

class ImageProcessor : public rclcpp::Node {
public:
    ImageProcessor() : Node("image_processor") {
        // Load the model
        model_ = cv::dnn::readNetFromONNX(model_path);

        // Subscribe to the compressed image topic
        subscription_ = this->create_subscription<sensor_msgs::msg::CompressedImage>(
            "/zed/zed_node/left/image_rect_color/compressed",
            10,
            std::bind(&ImageProcessor::processImage, this, std::placeholders::_1));

        // Create the publisher for detected signs
        publisher_ = this->create_publisher<team_interfaces::msg::Signs>("detected_signs", 10);

        cross_parking_ = false;
        pit_in_ = false;
        pit_out_ = false;
        overtaking_ = false;
    }

private:
    void processImage(const sensor_msgs::msg::CompressedImage::SharedPtr msg) {
        // Convert ROS Image message to OpenCV image
        cv_bridge::CvImagePtr cvImage = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);

        // Preprocess input
        int height = cvImage->image.rows;
        int width = cvImage->image.cols;
        int length = std::max(height, width);
        cv::Mat image = cv::Mat::zeros(length, length, CV_8UC3);
        cvImage->image.copyTo(image(cv::Rect(0, 0, width, height)));
        cv::Mat blob = cv::dnn::blobFromImage(image, 1.0 / 255, cv::Size(640, 640), cv::Scalar(0, 0, 0), true, false);

        // Set input to the model
        model_.setInput(blob);

        // Run inference
        cv::Mat outputs = model_.forward();

        // Preprocess output
        outputs = outputs.reshape(1, outputs.total() / 6);

        std::vector<cv::Rect> boxes;
        std::vector<float> scores;
        std::vector<int> class_ids;

        for (int i = 0; i < outputs.rows; ++i) {
            cv::Mat scores_class = outputs.row(i).colRange(4, outputs.cols);
            double min_score, max_score;
            cv::Point min_loc, max_loc;
            cv::minMaxLoc(scores_class, &min_score, &max_score, &min_loc, &max_loc);

            if (max_score >= 0.25) {
                cv::Rect box(outputs.at<float>(i, 0) - (0.5 * outputs.at<float>(i, 2)),
                             outputs.at<float>(i, 1) - (0.5 * outputs.at<float>(i, 3)),
                             outputs.at<float>(i, 2), outputs.at<float>(i, 3));
                boxes.push_back(box);
                scores.push_back(static_cast<float>(max_score));
                class_ids.push_back(max_loc.x);
            }
        }

        // Publish the detections
        auto result_msg = std::make_shared<team_interfaces::msg::Signs>();
        int max_class_id = 6;

        if (!scores.empty()) {
            int max_score_index = std::distance(scores.begin(), std::max_element(scores.begin(), scores.end()));
            max_class_id = class_ids[max_score_index];
            result_msg->sign_detected = true;
        } else {
            result_msg->sign_detected = false;
        }

        if (max_class_id == 0) {
            cross_parking_ = true;
        } else if (max_class_id == 1) {
            overtaking_ = true;
        } else if (max_class_id == 2) {
            overtaking_ = false;
        } else if (max_class_id == 3) {
            cross_parking_ = false;
        } else if (max_class_id == 4) {
            pit_in_ = true;
            pit_out_ = false;
        } else if (max_class_id == 5) {
            pit_out_ = true;
            pit_in_ = false;
        }

        result_msg->cross_parking = cross_parking_;
        result_msg->overtaking = overtaking_;
        result_msg->pit_in = pit_in_;
        result_msg->pit_out = pit_out_;

        publisher_->publish(result_msg);
    }

    cv::dnn::Net model_;
    rclcpp::Subscription<sensor_msgs::msg::CompressedImage>::SharedPtr subscription_;
    rclcpp::Publisher<team_interfaces::msg::Signs>::SharedPtr publisher_;

    bool cross_parking_;
    bool pit_in_;
    bool pit_out_;
    bool overtaking_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    auto image_processor = std::make_shared<ImageProcessor>();
    rclcpp::spin(image_processor);
    rclcpp::shutdown();
    return 0;
}

File: src/traffic_control_system_detection/scripts/image_processor_only_opencv.py
#!/usr/bin/env python3

import cv2
#import cv2.dnn.cuda as cuda_dnn #*1 for gpu use (open cv with cuda support necessary)
import numpy as np
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import CompressedImage
from cv_bridge import CvBridge
from team_interfaces.msg import Signs
import os
from ament_index_python.packages import get_package_share_directory

#cuda_dnn.dnn_cuda.init_device() #*1 for gpu use (open cv with cuda support necessary)
CLASSES = ['cross_parking','overtaking_allowed','overtaking_forbidden','parallel_parking','pit_in','pit_out']

class ImageProcessor(Node):
    def __init__(self):
        super().__init__('image_processor')
        #set up model
        model_path = os.path.join(get_package_share_directory('sign_detection'), 'model_files', 'best.onnx')
        self.model = cv2.dnn.readNetFromONNX(model_path)    #model = cuda_dnn.readNetFromONNX(model_path) #*1
        self.subscription = self.create_subscription(CompressedImage,'/zed/zed_node/left/image_rect_color/compressed',self.process_image,10)
        self.publisher = self.create_publisher(Signs,'detected_signs',10)
        self.cv_bridge = CvBridge()
        self.cross_parking = False
        self.pit_in = False
        self.pit_out = False
        self.overtaking = False


    def process_image(self, msg):
        # Convert ROS Image message to OpenCV image
        cv_image = self.cv_bridge.compressed_imgmsg_to_cv2(msg, desired_encoding='bgr8')

        # Preprocess input
        [height, width, _] = cv_image.shape
        length = max((height, width))
        image = np.zeros((length, length, 3), np.uint8)
        image[0:height, 0:width] = cv_image
        blob = cv2.dnn.blobFromImage(image, scalefactor=1 / 255, size=(640, 640), swapRB=True)    #blob = cuda_dnn.blobFromImage(image, scalefactor=1 / 255, size=(640, 640), swapRB=True) #*1
        self.model.setInput(blob)

        # Run inference
        outputs = self.model.forward()

        # Preprocess output
        outputs = np.array([cv2.transpose(outputs[0])])    #outputs = np.array([cuda_dnn.imagesFromBlob(outputs)[0]]) #*1
        rows = outputs.shape[1]

        boxes = []
        scores = []
        class_ids = []

        for i in range(rows):
            classes_scores = outputs[0][i][4:]
            (minScore, maxScore, minClassLoc, (x, maxClassIndex)) = cv2.minMaxLoc(classes_scores)
            if maxScore >= 0.25:
                box = [
                    outputs[0][i][0] - (0.5 * outputs[0][i][2]), outputs[0][i][1] - (0.5 * outputs[0][i][3]),
                    outputs[0][i][2], outputs[0][i][3]]
                boxes.append(box)
                scores.append(maxScore)
                class_ids.append(maxClassIndex)
 


        # Publish the detections as ROS String message
        self.result_msg = Signs()
        max_class_id = 6
        if scores:
            max_score_index = np.argmax(scores)
            #max_score = scores[max_score_index]
            #max_box = boxes[max_score_index]
            max_class_id = class_ids[max_score_index]
            self.result_msg.sign_detected = True
        else:
            self.result_msg.sign_detected = False

        if max_class_id == 0:
            self.cross_parking = True
        elif max_class_id == 1:
            self.overtaking = True
        elif max_class_id == 2:
            self.overtaking = False
        elif max_class_id == 3:
            self.cross_parking = False
        elif max_class_id == 4:
            self.pit_in = True
            self.pit_out = False
        elif max_class_id == 5:
            self.pit_out = True
            self.pit_in = False

        self.result_msg.cross_parking = self.cross_parking
        self.result_msg.overtaking = self.overtaking
        self.result_msg.pit_in = self.pit_in
        self.result_msg.pit_out = self.pit_out

        # Fix the line below to use self.result_msg
        self.publisher.publish(self.result_msg)

            



def main(args=None):
    rclpy.init(args=args)
    image_processor = ImageProcessor()
    rclpy.spin(image_processor)
    image_processor.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()